File: cpp/protein-translation/protein_translation.cpp
#include "protein_translation.h"

using namespace std;

namespace protein_translation
{
    vector<string> proteins(const string &proteins)
    {
        vector<string> translated;

        for (size_t i = 0; i < proteins.size() - 2; i += 3)
        {
            string protein = proteins.substr(i, 3);

            if (TRANSLATIONS.count(protein))
            {
                if (TRANSLATIONS.at(protein) == "STOP")
                {
                    break;
                }

                translated.push_back(TRANSLATIONS.at(protein));
            }
        }

        return translated;
    }
} // namespace protein_translation

File: python/protein-translation/protein_translation.py
from typing import List


def proteins(strand: str) -> List[str]:
    PROTEIN_TRANSLATION = {
        "AUG": "Methionine",
        "UUU": "Phenylalanine",
        "UUC": "Phenylalanine",
        "UUA": "Leucine",
        "UUG": "Leucine",
        "UCU": "Serine",
        "UCC": "Serine",
        "UCA": "Serine",
        "UCG": "Serine",
        "UAU": "Tyrosine",
        "UAC": "Tyrosine",
        "UGU": "Cysteine",
        "UGC": "Cysteine",
        "UGG": "Tryptophan",
        "UAA": "STOP",
        "UAG": "STOP",
        "UGA": "STOP",
    }

    proteins: List[str] = []

    for i in range(0, len(strand), 3):
        rna_sequence: str = strand[i : i + 3]

        if rna_sequence in PROTEIN_TRANSLATION.keys():
            if PROTEIN_TRANSLATION[rna_sequence] == "STOP":
                break

            proteins.append(PROTEIN_TRANSLATION[rna_sequence])

    return proteins
